include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/EdgeFunctions.h

```diff
@@ -472,6 +482,149 @@ template <typename AnalysisDomainTy> class EdgeFunctions {
+  operator=(const EdgeFunctionSingletonFactory &) = default;
+  EdgeFunctionSingletonFactory(EdgeFunctionSingletonFactory &&) noexcept =
+      default;
+  EdgeFunctionSingletonFactory &
+  operator=(EdgeFunctionSingletonFactory &&) noexcept = default;
+
+  virtual ~EdgeFunctionSingletonFactory() {
+    std::lock_guard<std::mutex> DataLock(getCacheData().DataMutex);
+  }
+
+  // Creates a new EdgeFunction of type EdgeFunctionType, reusing the previous
+  // allocation if an EdgeFunction with the same values was already created.
+  static inline std::shared_ptr<EdgeFunctionType>
+  createEdgeFunction(CtorArgT K) {
+    std::lock_guard<std::mutex> DataLock(getCacheData().DataMutex);
+
+    auto &Storage = getCacheData().Storage;
+    auto SearchVal = Storage.find(K);
+    if (SearchVal != Storage.end() && !SearchVal->second.expired()) {
+      return SearchVal->second.lock();
+    } else {
+      auto NewEdgeFunc = std::make_shared<EdgeFunctionType>(K);
+      Storage[K] = NewEdgeFunc;
+      return NewEdgeFunc;
+    }
+  }
+
+  // Initialize a cleaner thread to automatically remove unused/expired
+  // EdgeFunctions. The thread is only started the first time this function is
+  // called.
+  static inline void initEdgeFunctionCleaner() { getCleaner(); }
+
+  static inline void stopEdgeFunctionCleaner() { getCleaner().stop(); }
+
+  // Clean all unused/expired EdgeFunctions from the internal storage.
+  static inline void cleanExpiredEdgeFunctions() {
+    cleanExpiredMapEntries(getCacheData());
+  }
+
+  LLVM_DUMP_METHOD
+  static void dump(bool PrintElements = false) {
+    std::lock_guard<std::mutex> DataLock(getCacheData().DataMutex);
+
+    std::cout << "Elements in cache: " << getCacheData().Storage.size();
+
+    if (PrintElements) {
+      std::cout << "\n";
+      for (auto &KVPair : getCacheData().Storage) {
+        std::cout << "(" << KVPair.first << ") -> " << std::boolalpha
+                  << KVPair.second.expired() << std::endl;
+      }
+    }
+    std::cout << std::endl;
+  }
+
+private:
+  struct EFStorageData {
+    std::map<CtorArgT, std::weak_ptr<EdgeFunctionType>> Storage{};
+    std::mutex DataMutex;
+  };
+
+  static inline EFStorageData &getCacheData() {
+    static EFStorageData StoredData{};
+    return StoredData;
+  }
+
+  static void cleanExpiredMapEntries(EFStorageData &CData) {
+    std::lock_guard<std::mutex> DataLock(CData.DataMutex);
+
+    auto &Storage = CData.Storage;
+    for (auto Iter = Storage.begin(); Iter != Storage.end();) {
+      if (Iter->second.expired()) {
+        Iter = Storage.erase(Iter);
+      } else {
+        ++Iter;
+      }
+    }
+  }
+
+  class EdgeFactoryStorageCleaner {
+  public:
+    EdgeFactoryStorageCleaner()
+        : CleanerThread(&EdgeFactoryStorageCleaner::runCleaner, this,
+                        std::reference_wrapper<EFStorageData>(
+                            EdgeFunctionSingletonFactory::getCacheData())) {
+      pthread_setname_np(CleanerThread.native_handle(),
```

Comment:
the windows counterpart could be https://stackoverflow.com/a/23899379
honestly, that windows code looks pretty ugly.

Reply:
An easy alternative for windows could also be to just remove the thread naming. Makes it a bit less convenient for windows users but saves us some trouble.

Reply:
According to the pthread_setname_np man page it states
"""
The pthread_setname_np() function can be used to  set  a
       unique  name  for  a  thread,  which can be useful for debugging multi‐
       threaded applications.  The thread name  is  a  meaningful  C  language
       string, whose length is restricted to 16 characters, including the ter‐
       minating null byte ('\0').
"""
But anyways, I think we could get rid of thread naming.

Reply:
You want to fully remove it? It's quite handy for unix envs.

Reply:
I added ifdefs so only unix systems use thread naming